### PR TITLE
[WIP] passing custom dialer into spdy roundtripper

### LINKED ISSUE #71808: upgraded requests to aggregated API servers do not work when using ssh tunnels
**What happened**:

upgrade requests to aggregated APIs (like websocket-based watch requests) on a kube-apiserver using --ssh-user hang

**What you expected to happen**:

Websocket-based watch requests work properly

**How to reproduce it (as minimally and precisely as possible)**:

1. Start a kube-apiserver configured to use a node dialer to reach services/pods (e.g. with `--ssh-user`)
2. Ran an aggregated API server (like the sample-apiserver) on a node other than the kube-apiserver
3. Start a websocket-based watch against the aggregated API

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by caesarxuchao**
/kind bug
/sig api-machinery
/release-note none

Fixes #71808.

I'm not satisfied with the signature changes yet, but want to publish it asap to get some early feedback.

/assign @cheftako

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 1]
```diff
@@ -162,6 +162,12 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 
 	// we need to wrap the roundtripper in another roundtripper which will apply the front proxy headers
+	//
+	// Question(caesarxuchao): if the req contains upgrade
```
Comment (caesarxuchao) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:166]: I don't understand why aggregator needs an spdy.Roundtripper. I think we can just create a regular upgrade-aware round tripper here. If that's true, we can simplify this code a lot but that will also be a bigger change.

The current PR is a more targeted fix. Let's test it out first.

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 2]
```diff
@@ -231,12 +236,47 @@ func (s *SpdyRoundTripper) dial(req *http.Request) (net.Conn, error) {
 	return tlsConn, nil
 }
 
+func (s *SpdyRoundTripper) tlsDialWithDialFn(ctx context.Context, dialAddr string) (*tls.Conn, error) {
+	rawConn, err := s.DialFn(ctx, "tcp", dialAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	colonPos := strings.LastIndex(dialAddr, ":")
```
Comment (cheftako) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:245]: What if dialAddr is a IPv6 address with no port?

Comment (caesarxuchao) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:245]: Good point. I'll switch to net.SplitHostPort.

This was copied from the tls stand lib (https://cs.opensource.google/go/go/+/refs/tags/go1.17:src/crypto/tls/tls.go;l=138). Curious why it hasn't had any problem yet.

Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:245]: is it possible to pass an address without port?

Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:245]: it seem you can skip the address for tcp, but not the port https://cs.opensource.google/go/go/+/refs/tags/go1.17:src/net/dial.go;l=278-297

**Comment by liggitt** [comment]
closing PR from a k/k branch, can move to a personal fork branch

**Comment by spiffxp** [comment]
GitHub Admin super powers used: I had to manually disable branch protection for this branch to be able to delete it from the repo

**Comment by spiffxp** [comment]
I'm curious why we didn't get a slack notification about this branch getting created

**Pull request closed without merging**